modules/apps/frontend-data-set/frontend-data-set-impl/src/main/java/com/liferay/frontend/data/set/internal/url/FDSAPIURLBuilderImpl.java

```diff
@@ -0,0 +1,150 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2025 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.frontend.data.set.internal.url;
+
+import com.liferay.frontend.data.set.url.FDSAPIURLBuilder;
+import com.liferay.frontend.data.set.url.FDSAPIURLResolver;
+import com.liferay.frontend.data.set.url.FDSAPIURLResolverRegistry;
+import com.liferay.petra.string.CharPool;
+import com.liferay.petra.string.StringBundler;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.portal.kernel.util.WebKeys;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * @author Daniel Sanz
+ */
+public class FDSAPIURLBuilderImpl implements FDSAPIURLBuilder {
+
+	public FDSAPIURLBuilderImpl(
+		String restApplication, String restEndpoint, String restSchema,
+		FDSAPIURLResolverRegistry fdsAPIURLResolverRegistry,
+		HttpServletRequest httpServletRequest) {
+
+		_restApplication = restApplication;
+		_restEndpoint = restEndpoint;
+		_restSchema = restSchema;
+		_fdsAPIURLResolverRegistry = fdsAPIURLResolverRegistry;
+		_httpServletRequest = httpServletRequest;
+	}
+
+	@Override
+	public FDSAPIURLBuilder addParameter(String name, String value) {
+		if (Validator.isNotNull(name) && Validator.isNotNull(value)) {
+			_queryStringItems.add(name + CharPool.EQUAL + value);
+		}
+
+		return this;
+	}
+
+	@Override
+	public FDSAPIURLBuilder addQueryString(String queryString) {
+		if (Validator.isNotNull(queryString)) {
+			_queryStringItems.add(queryString);
+		}
+
+		return this;
+	}
+
+	@Override
+	public String build() {
+		int size = 3 + (_queryStringItems.size() * 2);
+
+		StringBundler sb = new StringBundler(size);
+
+		sb.append("/o");
+
+		sb.append(
+			StringUtil.replaceLast(
+				_restApplication, "/v1.0", StringPool.BLANK));
```

Comment:
Pretty opinionated, but probably good enough for now. We'll need some `if`s if we want this to work with any endpoint, not just objects.

Reply:
yes, for the moment I've just carried over preexisting logic. I'm not particularly happy with this bit but we can move on

Reply:
Also, some objects are `v2.0`, like `data-engine`. We may want to have regex here for version.